src/object.ts

```diff
@@ -0,0 +1,538 @@
+import * as http from 'http';
+import * as request from 'request';
+import {
+    ApisApi,
+    HttpError,
+    ObjectSerializer,
+    V1APIResource,
+    V1APIResourceList,
+    V1DeleteOptions,
+    V1Status,
+} from './api';
+import { KubeConfig } from './config';
+import { KubernetesListObject, KubernetesObject } from './types';
+
+/** Union type of body types returned by KubernetesObjectApi. */
+type KubernetesObjectResponseBody =
+    | KubernetesObject
+    | KubernetesListObject<KubernetesObject>
+    | V1Status
+    | V1APIResourceList;
+
+/** Kubernetes API verbs. */
+type KubernetesApiAction = 'create' | 'delete' | 'patch' | 'read' | 'replace';
+
+/**
+ * Valid Content-Type header values for patch operations.  See
+ * https://kubernetes.io/docs/tasks/run-application/update-api-object-kubectl-patch/
+ * for details.
+ */
+enum KubernetesPatchStrategies {
+    /** Diff-like JSON format. */
+    JsonPatch = 'application/json-patch+json',
+    /** Simple merge. */
+    MergePatch = 'application/merge-patch+json',
+    /** Merge with different strategies depending on field metadata. */
+    StrategicMergePatch = 'application/strategic-merge-patch+json',
+}
+
+/**
+ * Dynamically construct Kubernetes API request URIs so client does not have to know what type of object it is acting
+ * on.
+ */
+export class KubernetesObjectApi extends ApisApi {
+    /**
+     * Create a KubernetesObjectApi object from the provided KubeConfig.  This method should be used rather than
+     * [[KubeConfig.makeApiClient]] so we can properly determine the default namespace if one is provided by the current
+     * context.
+     *
+     * @param kc Valid Kubernetes config
+     * @return Properly instantiated [[KubernetesObjectApi]] object
+     */
+    public static makeApiClient(kc: KubeConfig): KubernetesObjectApi {
+        const client = kc.makeApiClient(KubernetesObjectApi);
+        if (kc.currentContext) {
+            const currentContext = kc.getContextObject(kc.currentContext);
+            if (currentContext && currentContext.namespace) {
+                client.defaultNamespace = currentContext.namespace;
+            }
+        }
+        return client;
+    }
+
+    /**
+     * Return whether the name of the resource should be appended to the API URI path.  When creating resources, it is
+     * not appended.
+     *
+     * @param action API action, see [[K8sApiAction]]
+     * @return true if name should be appended to URI
+     */
+    protected static appendName(action: KubernetesApiAction): boolean {
```

Comment:
This method feels unnecessary since it's just a single boolean compare. Just use the comparison everywhere.

Reply:
I'd prefer to keep this as the name of the method serves as documentation of the intent. The logic expressed here:

```typescript
        if (KubernetesObjectApi.appendName(action)) {
            if (!spec.metadata.name) {
                throw new Error('Required spec property name is not set');
            }
            parts.push(encodeURIComponent(String(spec.metadata.name)));
        }
```

is more clear than this:

```
        if (action !== 'create') {
            if (!spec.metadata.name) {
                throw new Error('Required spec property name is not set');
            }
            parts.push(encodeURIComponent(String(spec.metadata.name)));
        }
```

Although admittedly, the utility of the method was higher when `list` was supported and the body of the method looked `!(action === 'create' || action === 'list')`.

Reply:
Although perhaps the error being thrown communicates enough.

Reply:
I actually feel like the one with `'create'` is clearer. Otherwise I have to jump into the method to understand why/when you need to append the name.

Reply:
I did change it as you requested.